### ISSUE TITLE
Copy particles to scratch in float16 mode

### ISSUE THREAD

**Issue by rdrighetto** [3dem/relion#959]
Hi,

When reconstructing pseudo-subtomograms in RELION-4.0 we have the option to write the output in `float16`, which is great for saving space.

However, in `relion_refine` when we choose to copy particles to a scratch disk, the particles are copied as `float32`, meaning they take double the space in the scratch disk, and worse, slows down the copy process significantly.

Would it be possible to modify `src/exp_model.cpp` to write particles to scratch in the same mode they have originally (i.e. `float16` in this case)?

Unfortunately my C++ skills are almost zero, so I ask for help from the experts :-)

Thanks a lot!

**Comment by biochem-fan** [comment]
Indeed we are aware of this but this requires a non-trivial amount of code refactoring.
Note that the input particles can be a mixture of different box sizes, pixel sizes and/or float16/32.

Ideally, the refactoring should also implement parallel copy and grouped copy (read continuous particles in one go) to reduce IOPS.

**Comment by biochem-fan** [comment]
If you accept writing in float16, even when the input is float32, you can hard-code the float16 flag to your local installation.

It is the fifth argument to `Image.write()`. See https://github.com/3dem/relion/blob/master/src/preprocessing.cpp#L1193 for example.

**Comment by rdrighetto** [comment]
Thanks for the tip, @biochem-fan !

Actually, I did something like that indeed. I modified `exp_model.cpp` to force writing pseudo-subtomos and their weights to scratch as `float16` (available here: https://github.com/rdrighetto/relion/commit/21fe7b81bcf65bf08eb919e488b5a9e3d5405346). However, copying to scratch performance was basically the same as the vanilla code, taking ~3h30min to copy ~420k particles with a small box size of 32^3 voxels.
I guess the small box size in this case, regardless of `float16` or `float32`, pales in comparison to the number of particles, which seems to dominate the process. To confirm this, I'm also copying particles to scratch manually, but I suspect it will take roughly the same time (already copying for > 2h40min and counting).

Any tips on how to optimize scratch performance with such a large number of subtomograms?

Thanks again!

**Comment by biochem-fan** [comment]
Does tomography write one file for each particle? Then it is terribly bad.
Most file systems, especially distributed file systems, are bad at dealing with huge number of files.
You have to process file system metadata to find where each file is physically located.
You are limited not by bytes per second (throughput) but IO operations per second (IOPS).

Copying in parallel might help a bit, but a fundamental solution is to store multiple particles in a stack.

**Comment by rdrighetto** [comment]
Yes, that is exactly the problem: one file for each subtomogram, plus another file for their respective CTF weights.

Does RELION support MRC stacks of 3D objects? Not even sure if the MRC format itself supports that.
I know for example the IMAGIC format supports it.

Thanks!

**Comment by biochem-fan** [comment]
No, the MRC format has only 3 axes.
https://www.ccpem.ac.uk/mrc_format/mrc2014.php
